code/modules/mob/living/inhand_holder.dm

```diff
@@ -5,80 +5,141 @@
+	..()
+	if(isturf(loc))//don't release on soft-drops
+		release()
 
 /obj/item/clothing/head/mob_holder/proc/release(del_on_release = TRUE)
-	if(!held_mob)
-		if(del_on_release && !destroying)
-			qdel(src)
-		return FALSE
 	if(isliving(loc))
 		var/mob/living/L = loc
-		to_chat(L, "<span class='warning'>[held_mob] wriggles free!</span>")
 		L.dropItemToGround(src)
-	held_mob.forceMove(get_turf(held_mob))
-	held_mob.reset_perspective()
-	held_mob.setDir(SOUTH)
-	held_mob.visible_message("<span class='warning'>[held_mob] uncurls!</span>")
-	held_mob = null
-	if(del_on_release && !destroying)
+	if(held_mob)
+		var/mob/living/m = held_mob
+		m.forceMove(get_turf(m))
+		m.reset_perspective()
+		m.setDir(SOUTH)
+		held_mob = null
+	if(del_on_release)
 		qdel(src)
-	return TRUE
 
 /obj/item/clothing/head/mob_holder/relaymove(mob/user)
 	release()
 
 /obj/item/clothing/head/mob_holder/container_resist()
 	release()
 
-/obj/item/clothing/head/mob_holder/drone/deposit(mob/living/L)
-	. = ..()
-	if(!isdrone(L))
-		qdel(src)
-	name = "drone (hiding)"
-	desc = "This drone is scared and has curled up into a ball!"
-
-/obj/item/clothing/head/mob_holder/drone/update_visuals(mob/living/L)
-	var/mob/living/simple_animal/drone/D = L
-	if(!D)
-		return ..()
-	icon = 'icons/mob/drone.dmi'
-	icon_state = "[D.visualAppearence]_hat"
+/mob/living/proc/mob_pickup(mob/living/L)
+	var/obj/item/clothing/head/mob_holder/holder = generate_mob_holder()
+	if(!holder) return
+	drop_all_held_items()
+	L.put_in_hands(holder)
+	return
+
+/mob/living/proc/generate_mob_holder()
+	..()
+	var/obj/item/clothing/head/mob_holder/holder = new(get_turf(src), src, (istext(can_be_held) ? can_be_held : ""), 'icons/mob/animals_held.dmi', 'icons/mob/animals_held_lh.dmi', 'icons/mob/animals_held_rh.dmi')
+	return holder
+
+/mob/living/simple_animal/drone/generate_mob_holder()
+	var/obj/item/clothing/head/mob_holder/holder = new(get_turf(src), src, "[visualAppearence]_hat", null, null, null, TRUE)
+	return holder
+
+/mob/living/carbon/monkey/generate_mob_holder()
+	var/obj/item/clothing/head/mob_holder/holder = new(get_turf(src), src, "monkey", 'icons/mob/animals_held.dmi', 'icons/mob/animals_held_lh.dmi', 'icons/mob/animals_held_rh.dmi', TRUE)
+	return holder
+
+/mob/living/simple_animal/mouse/generate_mob_holder()
+	var/obj/item/clothing/head/mob_holder/holder = new(get_turf(src), src, (istext(can_be_held) ? can_be_held : ""), 'icons/mob/animals_held.dmi', 'icons/mob/animals_held_lh.dmi', 'icons/mob/animals_held_rh.dmi')
+	holder.w_class = 1
+	return holder
+
+/mob/living/simple_animal/hostile/lizard/generate_mob_holder()
+	var/obj/item/clothing/head/mob_holder/holder = new(get_turf(src), src, "lizard", 'icons/mob/animals_held.dmi', 'icons/mob/animals_held_lh.dmi', 'icons/mob/animals_held_rh.dmi', TRUE)
+	return holder
+
+/mob/living/proc/mob_try_pickup(mob/living/user)
+	if(!ishuman(user) || !src.Adjacent(user) || user.incapacitated() || !can_be_held)

```

Comment:
no implied `src`. Here or anywhere else unless you ever change `src` or have poorly named variables.